dart/dynamics/IkFast.cpp

```diff
@@ -0,0 +1,329 @@
+
+#include "dart/external/ikfast/ikfast.h"
+#include "dart/dynamics/BodyNode.hpp"
+#include "dart/dynamics/DegreeOfFreedom.hpp"
+
+namespace dart {
+namespace dynamics {
+
+namespace {
+
+//==============================================================================
+void convertTransform(
+    const IkReal* eetrans, const IkReal* eerot, Eigen::Isometry3d& tf)
+{
+  tf.linear()(0, 0) = eerot[0*3+0];
+  tf.linear()(0, 1) = eerot[0*3+1];
+  tf.linear()(0, 2) = eerot[0*3+2];
+  tf.linear()(1, 0) = eerot[1*3+0];
+  tf.linear()(1, 1) = eerot[1*3+1];
+  tf.linear()(1, 2) = eerot[1*3+2];
+  tf.linear()(2, 0) = eerot[2*3+0];
+  tf.linear()(2, 1) = eerot[2*3+1];
+  tf.linear()(2, 2) = eerot[2*3+2];
+
+  tf.translation() << eetrans[0], eetrans[1], eetrans[2];
+}
+
+//==============================================================================
+void convertTransform(
+    const Eigen::Isometry3d& tf,
+    std::array<IkReal, 3>& eetrans,
+    std::array<IkReal, 9>& eerot)
+{
+  eerot[0*3+0] = tf.linear()(0, 0);
+  eerot[0*3+1] = tf.linear()(0, 1);
+  eerot[0*3+2] = tf.linear()(0, 2);
+  eerot[1*3+0] = tf.linear()(1, 0);
+  eerot[1*3+1] = tf.linear()(1, 1);
+  eerot[1*3+2] = tf.linear()(1, 2);
+  eerot[2*3+0] = tf.linear()(2, 0);
+  eerot[2*3+1] = tf.linear()(2, 1);
+  eerot[2*3+2] = tf.linear()(2, 2);
+
+  eetrans[0] = tf.translation()[0];
+  eetrans[1] = tf.translation()[1];
+  eetrans[2] = tf.translation()[2];
+}
+
+//==============================================================================
+bool isFreeJoint(int numFreeParams, const int* freeParams, int index)
+{
+  for (auto i = 0; i < numFreeParams; ++i)
+  {
+    if (index == freeParams[i])
+      return true;
+  }
+
+  return false;
+}
+
+//==============================================================================
+void convertIkSolution(
+    const IkFast* ikfast,
+    int numJoints,
+    int numFreeParameters,
+    int* freeParameters,
+    const ikfast::IkSolutionBase<IkReal>& ikfastSolution,
+    InverseKinematics::Analytical::Solution& solution)
+{
+  const auto ik = ikfast->getIK();
+  const auto dofIndices = ikfast->getDofs();
+  const auto skel = ik->getNode()->getSkeleton();
+
+  std::vector<IkReal> solutionValues(numJoints);
+  std::vector<IkReal> freeValues(numFreeParameters);
+
+  ikfastSolution.GetSolution(solutionValues, freeValues);
+
+  bool limitViolated = false;
+
+  auto index = 0u;
+  solution.mConfig.resize(dofIndices.size());
+  for (auto i = 0u; i < solutionValues.size(); ++i)
+  {
+    if (isFreeJoint(numFreeParameters, freeParameters, i))
+      continue;
+
+    solution.mConfig[index] = solutionValues[i];
+
+    const auto dofIndex = dofIndices[index];
+
+    if (solutionValues[i] < skel->getDof(dofIndex)->getPositionLowerLimit())
+    {
+      limitViolated = true;
+      break;
+    }
+
+    if (solutionValues[i] > skel->getDof(dofIndex)->getPositionUpperLimit())
+    {
+      limitViolated = true;
+      break;
+    }
+
+    index++;
+  }
+
```

Comment:
This class - or the `Analytic` base class - should support a `bool` flag that toggles error checking. It's inexpensive to verify that running the IK solver's solution through DART's forward kinematics returns the expected `BodyNode` pose.

This is a very important check - to the extent that we may want to enable it by default - because it catches to broad classes of errors: (1) using an analytic IK solution for a slightly different model and (2) bugs in the analytic IK solution. OpenRAVE avoids the former by hashing the model - which is not really possible here. I have also been burned by the latter due to bugs in IKFast.